pkg/starknet/state.go

```diff
@@ -0,0 +1,635 @@
+				)
+
+				stateDiff := parsePages(pages)
+
+				// Update state
+				latestBlockSynced = s.updateAndCommitState(stateDiff, fact.StateRoot, fact.SequenceNumber)
+
+				// update services
+				go s.updateServices(*stateDiff, "", strconv.FormatUint(fact.SequenceNumber, 10))
+
+				isNoErr := s.facts.Remove(strconv.FormatUint(latestBlockSynced-1, 10))
+				if !isNoErr {
+					return
+				}
+			}
+		}
+	}()
+
+	for l := range event {
+		// Process GpsStatementVerifier contract
+		factHash, ok := l.Event["factHash"]
+		pagesHashes, ok1 := l.Event["pagesHashes"]
+
+		if ok && ok1 {
+			b := make([]byte, 0)
+			for _, v := range factHash.([32]byte) {
+				b = append(b, v)
+			}
+			value := starknetTypes.PagesHash{Bytes: pagesHashes.([][32]byte)}
+			s.gpsVerifier.Add(common.BytesToHash(b).Hex(), value)
+		}
+		// Process MemoryPageFactRegistry contract
+		if memoryHash, ok := l.Event["memoryHash"]; ok {
+			key := common.BytesToHash(memoryHash.(*big.Int).Bytes()).Hex()
+			value := starknetTypes.TransactionHash{Hash: l.TransactionHash}
+			s.memoryPageHash.Add(key, value)
+		}
+		// Process Starknet logs
+		if fact, ok := l.Event["stateTransitionFact"]; ok {
+
+			b := make([]byte, 0)
+			for _, v := range fact.([32]byte) {
+				b = append(b, v)
+			}
+			contractAbi, _ := loadAbiOfContract(abi.StarknetAbi)
+			starknetAddress := common.HexToAddress(contractAddresses.Starknet)
+
+			if err != nil {
+				log.Default.With("Error", err).
+					Info("Unable to get the Value of the latest fact synced")
+				return err
+			}
+
+			blockNumber := new(big.Int).SetUint64(l.Block)
+			query := ethereum.FilterQuery{
+				FromBlock: blockNumber,
+				ToBlock:   blockNumber,
+				Addresses: []common.Address{starknetAddress},
+				Topics:    [][]common.Hash{{crypto.Keccak256Hash([]byte(contractAbi.Events["LogStateUpdate"].Sig))}},
+			}
+
+			starknetLogs, err := s.ethereumClient.FilterLogs(context.Background(), query)
+			if err != nil {
+				log.Default.With("Error", err, "Initial block", l.Block, "End block", l.Block+1).
+					Info("Couldn't get logs")
+			}
+			fullFact, err := getFactInfo(starknetLogs, contractAbi, common.BytesToHash(b).Hex(), latestBlockSaved, l.TransactionHash)
+			if err != nil {
+				continue
+			}
+
+			// Safe Fact for block x
+			s.facts.Add(strconv.FormatUint(latestBlockSaved, 10), fullFact)
+			latestBlockSaved++
+		}
+	}
+	return errors.New("events channel closed")
+}
+
+// updateAndCommitState applies `stateDiff` to the local state and
+// commits the changes to the database.
+// notest
+func (s *Synchronizer) updateAndCommitState(
+	stateDiff *starknetTypes.StateDiff,
+	newRoot string,
+	sequenceNumber uint64,
+) uint64 {
+	txn := s.transactioner.Begin()
+	hashService := services.GetContractHashService()
+	if hashService == nil {
+		log.Default.Panic("Contract hash service is unavailable")
+	}
+	_, err := updateState(txn, services.GetContractHashService(), stateDiff, newRoot, sequenceNumber)
```

Comment:
Use the `hashService` variable created earlier in the function.

Reply:
Fixed